File: PracticasProgramacion/Practica_15/cpp/Recursividad.cpp
#include "Recursividad.h"
#include <stdexcept>

long long Recursividad::factorial(int n) const {
    if (n < 0) throw std::runtime_error("Factorial de negativo no definido");
    if (n == 0 || n == 1) return 1;
    return n * factorial(n - 1);
}

long long Recursividad::fibonacci(int n) const {
    if (n < 0) throw std::runtime_error("Fibonacci de negativo no definido");
    if (n == 0) return 0;
    if (n == 1) return 1;
    return fibonacci(n - 1) + fibonacci(n - 2);
}

File: PracticasProgramacion/Practica_15/python/recursividad.py
"""
Recursividad: factorial y fibonacci con llamadas recursivas.
Caso base: n<=1 para factorial; n in (0,1) para fibonacci.
"""


class Recursividad:
    def factorial(self, n: int) -> int:
        if n < 0:
            raise ValueError("Factorial de negativo no definido")
        if n in (0, 1):
            return 1
        return n * self.factorial(n - 1)

    def fibonacci(self, n: int) -> int:
        if n < 0:
            raise ValueError("Fibonacci de negativo no definido")
        if n == 0:
            return 0
        if n == 1:
            return 1
        return self.fibonacci(n - 1) + self.fibonacci(n - 2)
